### ISSUE TITLE
Investigate possible ring modulation aliasing on keys emulator

### ISSUE THREAD

**Issue by waterjump** [waterjump/volca-share#43]
**TLDR:** Ring mod causes frequencies above the audible frequency range that could result in audible harmonics because the sample rate is not twice as high as them.
**Impact:** Medium to low
**Feasibility of fix**: Uncertain give limitations of JS sample rate control.

Reddit user Lopiano [suggested in a comment](https://www.reddit.com/r/synthesizers/comments/1f0eyzb/comment/ljtzd4g/) there's a potential for aliasing when ring mod is in use.  I looked into it and I think it's possible but I have yet to observe or intentionally create aliasing using the keys emulator.

**Description of problem**
To over-simplify, when two waves are a part of a ring modulation algorithm, harmonics are added at the sum and the difference of the carrier wave and modulating wave.  For example a 500 Hz carrier wave modulated by a 700Hz wave will result in harmonics at 1200 Hz and 200 Hz.  In some scenarios, the harmonics created by modulation might exceed 22050 Hz, which is the Nyquist frequency of the standard 44100 sample rate of most Web Audio environments.  When frequencies above the Nyquist frequency are sampled, they can result in aliasing, which would take the form of a harmonic at the value of 44100 minus the difference of the harmonic and the Nyquist frequency.  Eg. - A 43100 Hz wave sampled at 44100 Hz would produce an alias wave at 1000 Hz.

**What aliasing looks like in the keys emulator**
To start by only looking at fundamental frequencies of notes and ignoring pitch effects like detune, LFO etc, the three highest notes that can be played (without affecting pitch with detune, LFO etc) are A#8 with 7458 Hz, B8 with 7902 Hz, and C9 with 8372 Hz.  Since the lowest note is always the carrier, the sum and difference of A#8 and B8 are 15356 Hz and 444 Hz, neither of which would cause aliasing because they're below 22050 Hz.  But if that signal is modulated by C9, we could expect a harmonic at 23731, which would cause an alias harmonic at 20369.  I have confirmed this by analyzing the spectrum in Audacity:
<img width="1000" alt="Screen Shot 2024-08-25 at 8 26 04 PM" src="https://github.com/user-attachments/assets/ee58b15d-47b5-4f28-b1aa-8ef578c8ae93">

**Impact**
Although to me the aliasing in the test case in inaudible, and using notes that high is uncommon, the potential for aliasing is there.  Strictly from the listener's perspective, I can't say this alone is a disruptive enough issue to address, but I will keep my ear out for aliasing and try to come up with more impactful cases.  The nature of the emulator itself is non-scientific and my personal opinion is that the poly ring mod is mostly a wildcard to begin with, so the margin for allowing aliasing is relatively high.

**Potential solutions**
To minimize aliasing in ring modulation:
 - Oversampling: Increase the sampling rate so that the Nyquist frequency is higher, thereby reducing the likelihood of frequencies exceeding it.
 - Filtering: Apply low-pass filters to the signals before and after modulation to remove or reduce high-frequency components that could cause aliasing.
 - Bandlimiting: Limit the frequency range of the signals being modulated to ensure that their sum and difference frequencies stay within the Nyquist limit.

**Comment by waterjump** [comment]
Following up with another test case that is audible and easily observable in Audacity, playing the three highest notes in poly ring voice with LFO pitch intentsity turned all the way up shows clear aliasing.  [Patch to recreate](http://127.0.0.1:3000/keys/emulator?voice=120&detune=0&portamento=0&vco_eg_int=0&attack=0&decay_release=0&vcf_eg_int=0&peak=0&cutoff=127&sustain=127&lfo_rate=0&lfo_pitch_int=127&lfo_cutoff_int=0&delay_time=0&delay_feedback=0&octave=50&lfo_trigger_sync=false&lfo_shape=triangle)

<img width="782" alt="Screen Shot 2024-08-25 at 8 53 44 PM" src="https://github.com/user-attachments/assets/322465c6-103b-4641-aab3-8016251885bc">
